### PR TITLE
Stream CLI server command stderr instead of buffering

### PR DESCRIPTION
**Pull request by koesie10**
This streams the CLI server command stderr instead of buffering it and only printing it when the CLI command has completed. This helps users to know that something is actually happening for long-running commands. The full stderr will still be completed in an error when the command fails.

I think an easy way to test this is running a variant analysis. While the query is compiling, output should be streamed to the extension log output.

Closes https://github.com/github/vscode-codeql/issues/1459

## Checklist

- [ ] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [ ] Issues have been created for any UI or other user-facing changes made by this pull request.
- [ ] _[Maintainers only]_ If this pull request makes user-facing changes that require documentation changes, open a corresponding docs pull request in the [github/codeql](https://github.com/github/codeql/tree/main/docs/codeql/codeql-for-visual-studio-code) repo and add the `ready-for-doc-review` label there.

### CODE REVIEW

**Comment by koesie10** [comment]
> What's the reasoning behind using different logic for stdout and stderr? Shall we always stream?

The stdout usually doesn't contain output that users need to deal with: it contains the output that we consume from within the extension, for example a JSON object with resolved data. The stderr only contains output that is useful for users and is not parsed by us (except in case the command fails, in which case we sometimes try to extract the error message). So, we don't need to stream stdout because it isn't useful information for users (and usually also isn't streamed from the CLI side, only printed at the end of a command).

File: extensions/ql-vscode/src/codeql-cli/cli.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/codeql-cli/cli.ts:366]: Would it be simpler to store a single `currentLineBuffer: Buffer` instead of an array of buffers?

Currently, whenever we want to print the current line we have to concatenate all the buffers together, but this could be avoided if we stored the concatenated buffer. And I don't think it would make for extra work when adding a new data fragment. The current process is:
- add new fragment to `currentLineStderrBuffers` array
- Concatenate `currentLineStderrBuffers` array into `currentLineBuffer` and check if it contains an `EOL`
- After handling any newlines, set `currentLineStderrBuffers = [currentLineBuffer]`. Maintaining the original individual fragments isn't important and after handling any newlines that data is lost.

I think we could simplify so whenever any new fragment comes in we do `currentLineBuffer = Buffer.concat([currentLineBuffer, newData])` and then proceed as before, updating `currentLineBuffer` whenever we find a `EOL` and print a line.

File: extensions/ql-vscode/src/codeql-cli/cli.ts
Comment (robertbrignull) [inline extensions/ql-vscode/src/codeql-cli/cli.ts:448]: We check whether the buffer contains `eolBuffer` or calculate the index quite a lot here. Once first of all, then once as the loop condition, and then twice within the loop. I recognise it's not a very expensive operation of the buffer is in-memory and likely to be short, but I wonder if it's worth avoiding anyway.

I normally find it confusing the set a variable in the loop condition, but this could be a case where it works quite well. Then we calculate it only once per loop.

If you also go with my earlier suggestion of converting the array of buffers into a single buffer, the code could look like:
```suggestion
              let eolIndex;
              while (
                (eolIndex = currentLineStderrBuffer.indexOf(eolBuffer)) !== -1
              ) {
                const line = currentLineStderrBuffer.subarray(0, eolIndex);
                void this.logger.log(line.toString("utf-8"));

                currentLineStderrBuffer = currentLineStderrBuffer.subarray(
                  eolIndex + eolBuffer.length,
                );
              }
```

**Review by robertbrignull — APPROVED** [review]
I've tried it out and comparing it to `main` you can definitely tell that the output is snappier and more responsive than it was before.

I have a couple of small potential improvements for the code but they're not essential if others disagree, and the code works fine as it is.

I'm happy if you want to wait for a review from CodeQL Experiences or go ahead with just reviews from us.

File: extensions/ql-vscode/src/codeql-cli/cli.ts
Comment (koesie10) [reply extensions/ql-vscode/src/codeql-cli/cli.ts:448]: Thanks, I think that makes sense. I did find that because we're doing `subarray` here, we also need to copy the buffer to prevent memory leaks (because a subarray keeps a reference to the original buffer), but this was also present in my original code.

**Review by robertbrignull — APPROVED** [review]
Thanks for updating. The new changes LGTM too.

**Pull request merged**
